From 3ef732b55ff468bd60959df290c698e547d305d2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Mon, 16 Dec 2024 22:53:34 +0100
Subject: [PATCH 47/86] rpcrt4: Use no_sanitize_address in client_do_args.

/home/bernhard/data/entwicklung/2024/wine/wine/tools/runtest -q -P wine -T . -M msctf.dll -p dlls/msctf/tests/x86_64-windows/msctf_test.exe inputprocessor && touch dlls/msctf/tests/x86_64-windows/inputprocessor.ok
01bc:fixme:msctf:TextStoreACPSink_OnLockGranted OnLockGranted called for something other than an EditSession
=================================================================
==440==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x7ef431f3fb60 at pc 0x6ffff7cb66fe bp 0x7ef431f3e060 sp 0x7ef431f3e0a8
READ of size 8 at 0x7ef431f3fb60 thread T1
01d0:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff7cb66fd in client_do_args /home/bernhard/data/entwicklung/2024/wine\wine/dlls/rpcrt4/ndr_stubless.c:511:17
    #1 0x6ffff7cb958d in ndr_client_call /home/bernhard/data/entwicklung/2024/wine\wine/dlls/rpcrt4/ndr_stubless.c:789:13
    #2 0x6ffff7cb87ec in NdrpClientCall2 /home/bernhard/data/entwicklung/2024/wine\wine/dlls/rpcrt4/ndr_stubless.c:929:22
    #3 0x6ffff7d11d0b in call_stubless_func (C:\windows\system32\rpcrt4.dll+0x1800a1d0b)
    #4 0x6ffff7fe1fe1 in IRemUnknown_RemQueryInterface /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\dlls/combase\dcom.h:309:12
    #5 0x6ffff7fe1fe1 in ClientIdentity_QueryMultipleInterfaces /home/bernhard/data/entwicklung/2024/wine\wine/dlls/combase/marshal.c:1047:18
    #6 0x6ffff7fe0d13 in IMultiQI_QueryMultipleInterfaces /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include\objidl.h:1502:12
    #7 0x6ffff7fe0d13 in ClientIdentity_QueryInterface /home/bernhard/data/entwicklung/2024/wine\wine/dlls/combase/marshal.c:975:10
    #8 0x000140022aea in ITfThreadMgrEx_QueryInterface /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include\msctf.h:1311:12
    #9 0x000140022aea in test_MultiThreadApartment_Thread /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msctf/tests/inputprocessor.c:2595:10
    #10 0x6ffff902b17d in asan_thread_start(void*) /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_win.cpp:147:14
    #11 0x6ffffbdc4808 in BaseThreadInitThunk /usr/src/packages/BUILD\dlls/kernel32\thread.c:61:5
    #12 0x6ffffad1fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

Address 0x7ef431f3fb60 is located in stack of thread T1 at offset 0 in frame
    #0 0x6ffff7fe14bf in ClientIdentity_QueryMultipleInterfaces /home/bernhard/data/entwicklung/2024/wine\wine/dlls/combase/marshal.c:999

  This frame has 1 object(s):
    [32, 40) 'qiresults' (line 1001)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
Thread T1 created by T0 here:
    #0 0x6ffff902b096 in CreateThread /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_win.cpp:158:3
    #1 0x00014000a276 in test_MultiThreadApartment /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msctf/tests/inputprocessor.c:2618:14
    #2 0x00014000a276 in func_inputprocessor /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msctf/tests/inputprocessor.c:2659:9
    #3 0x000140024b33 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #4 0x00014002457b in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h
    #5 0x0001400264af in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #6 0x6ffffbdc4808 in BaseThreadInitThunk /usr/src/packages/BUILD\dlls/kernel32\thread.c:61:5
    #7 0x6ffffad1fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

SUMMARY: AddressSanitizer: stack-buffer-underflow /home/bernhard/data/entwicklung/2024/wine\wine/dlls/rpcrt4/ndr_stubless.c:511:17 in client_do_args
Shadow bytes around the buggy address:
  0x7ef431f3f880: 00 00 00 00 00 00 00 00 f2 f2 f2 f2 f2 f2 f2 f2
  0x7ef431f3f900: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ef431f3f980: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ef431f3fa00: f8 f8 f8 f8 f8 f8 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3
  0x7ef431f3fa80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x7ef431f3fb00: 00 00 00 00 00 00 00 00 00 00 00 00[f1]f1 f1 f1
  0x7ef431f3fb80: 00 f3 f3 f3 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ef431f3fc00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ef431f3fc80: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
  0x7ef431f3fd00: 00 00 00 f3 f3 f3 f3 f3 00 00 00 00 00 00 00 00
  0x7ef431f3fd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==440==ABORTING
make: *** [Makefile:225627: dlls/msctf/tests/x86_64-windows/inputprocessor.ok] Fehler 1
01d0:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
---
 dlls/rpcrt4/ndr_stubless.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/rpcrt4/ndr_stubless.c b/dlls/rpcrt4/ndr_stubless.c
index 8cb051443ae..8b54ed22b72 100644
--- a/dlls/rpcrt4/ndr_stubless.c
+++ b/dlls/rpcrt4/ndr_stubless.c
@@ -477,6 +477,7 @@ static size_t basetype_arg_size( unsigned char fc )
     }
 }
 
+__attribute__((no_sanitize_address))
 void client_do_args( PMIDL_STUB_MESSAGE pStubMsg, PFORMAT_STRING pFormat, enum stubless_phase phase,
                      BOOLEAN fpu_args, unsigned short number_of_params, unsigned char *pRetVal )
 {
-- 
2.47.1

